src/cgns_internals.c

```diff
@@ -121,6 +124,201 @@ int cgi_read()
     return CG_OK;
 }
 
+/* helpers to parse CGNSBase_t node */
+typedef enum {
+    LabelNull_t,
+    LabelFamily_t,
+    LabelReferenceState_t,
+    LabelGravity_t,
+    LabelAxisymmetry_t,
+    LabelRotatingCoordinates_t,
+    LabelConvergenceHistory_t,
+    LabelDescriptor_t,
+    LabelDataClass_t,
+    LabelDimensionalUnits_t,
+    LabelFlowEquationSet_t,
+    LabelIntegralData_t,
+    LabelSimulationType_t,
+    LabelBaseIterativeData_t,
+    LabelUserDefinedData_t,
+    LabelZone_t
+} BaseLabel_t;
+
+static int get_base_label_type_as_enum(const char* nodelabel) {
+    if (0 == strcmp(nodelabel, "Zone_t")) {
+        return LabelZone_t;
+    }
+    else if (0 == strcmp(nodelabel, "Family_t")) {
+        return LabelFamily_t;
+    }
+    else if (0 == strcmp(nodelabel, "ReferenceState_t")) {
+        return LabelReferenceState_t;
+    }
+    else if (0 == strcmp(nodelabel, "Gravity_t")) {
+        return LabelGravity_t;
+    }
+    else if (0 == strcmp(nodelabel, "Axisymmetry_t")) {
+        return LabelAxisymmetry_t;
+    }
+    else if (0 == strcmp(nodelabel, "RotatingCoordinates_t")) {
+        return LabelRotatingCoordinates_t;
+    }
+    else if (0 == strcmp(nodelabel, "ConvergenceHistory_t")) {
+        return LabelConvergenceHistory_t;
+    }
+    else if (0 == strcmp(nodelabel, "Descriptor_t")) {
+        return LabelDescriptor_t;
+    }
+    else if (0 == strcmp(nodelabel, "DataClass_t")) {
+        return LabelDataClass_t;
+    }
+    else if (0 == strcmp(nodelabel, "DimensionalUnits_t")) {
+        return LabelDimensionalUnits_t;
+    }
+    else if (0 == strcmp(nodelabel, "FlowEquationSet_t")) {
+        return LabelFlowEquationSet_t;
+    }
+    else if (0 == strcmp(nodelabel, "IntegralData_t")) {
+        return LabelIntegralData_t;
+    }
+    else if (0 == strcmp(nodelabel, "SimulationType_t")) {
+        return LabelSimulationType_t;
+    }
+    else if (0 == strcmp(nodelabel, "BaseIterativeData_t")) {
+        return LabelBaseIterativeData_t;
+    }
+    else if (0 == strcmp(nodelabel, "UserDefinedData_t")) {
+        return LabelUserDefinedData_t;
+    }
+    else {
+        return LabelNull_t;
+    }
+}
+
+typedef struct _childnode {
+    double id;
+    int type;
+    char_33 name;
+} _childnode_t;
+
+static int sort_base_children(const void* v1, const void* v2)
+{
+    _childnode_t* p1;
+    _childnode_t* p2;
+
+    p1 = (_childnode_t *) v1;
+    p2 = (_childnode_t *) v2;
+    return p1->type - p2->type;
+}
+
+static int sort_zone_names(const void* v1, const void* v2)
+{
+    _childnode_t* p1;
+    _childnode_t* p2;
+
+    p1 = (_childnode_t*)v1;
+    p2 = (_childnode_t*)v2;
+    return (strcmp(p1->name, p2->name));
+}
+
+#ifdef _WIN32
+#define CG_INLINE __inline
+#else
+#define CG_INLINE static inline
+#endif
+
+CG_INLINE int cgi_get_nodes_with_label(const _childnode_t *childlist, const int nchildren, BaseLabel_t label, int *nnodes, int *startpos)
```

Comment:
This return a section of the whold childlist that starts at starpos and ends at starpos+nnodes
Maybe returning a new array with only the node would be as efficient and remove the need of a sort by label